src/main/java/keqing/Keqing.java

```diff
@@ -0,0 +1,204 @@
+import keqing.tasks.Deadline;
+import keqing.tasks.Event;
+import keqing.tasks.Task;
+import keqing.tasks.ToDo;
+
+import java.util.Scanner;
+public class Keqing {
+
+    public static final String LINE = "____________________________________________________________\n";
+    public static void echo(Task[] tasks) {
+        System.out.println(LINE);
+        System.out.println("Got it. I've added this task:");
+        System.out.println("  added: " + tasks[Task.getTaskCount() - 1].toString());
+        System.out.println("Now you have " + Task.getTaskCount() + " tasks in your list.");
+        System.out.println(LINE);
+    }
+
+    public static void printTaskList(Task[] tasks) {
+        System.out.println(LINE);
+        if (Task.getTaskCount() == 0) {
+            System.out.println("The list is empty...!");
+        }
+        for (int i = 0; i < Task.getTaskCount(); i++) {
+            System.out.print((i + 1) + ".");
+            System.out.println(tasks[i].toString());
+        }
+        System.out.println(LINE);
+    }
+
+    public static void markTask(Task[] tasks, int currentID, boolean isDone) {
+        System.out.println(LINE);
+        if (currentID < 0 || currentID >= Task.getTaskCount()) {
+            System.out.println("Cannot find this task!");
+        }
+        else if (isDone) {
+            tasks[currentID].setDone();
+            System.out.println("Nice! I've marked this task as done:");
+            System.out.println("   " + tasks[currentID].toString());
+        }
+        else {
+            tasks[currentID].seUndone();
+            System.out.println("OK, I've marked this task as not done yet:");
+            System.out.println("   " + tasks[currentID].toString());
+        }
+        System.out.println(LINE);
+    }
+
+    public static void printMenu() {
+        System.out.println(LINE);
+        System.out.println("Try the following commands:");
+        System.out.println("1. list: show the list of tasks;");
+        System.out.println("2. todo: add a task without starting time/deadline to the list;");
+        System.out.println("3. deadline: add a task with deadline to the list;");
+        System.out.println("4. event: add a task with specific starting and ending time");
+        System.out.println("5, mark: mark a task as 'done' state;");
+        System.out.println("6. unmark: unmark a task from 'done' state;");
+        System.out.println(LINE);
+    }
+
+    public static void readToDo(Task[] tasks, String content) throws IllegalInputException {
+        if (content.equals("todo")) {
+            throw new IllegalInputException("Keqing doesn't understand what you actually want to do...");
+        }
+        else {
+            ToDo toDoTask = new ToDo(content, Task.getTaskCount());
+            tasks[Task.getTaskCount() - 1] = toDoTask;
+        }
+    }
+
+    public static void readDeadline(Task[] tasks, String content) throws IllegalInputException {
+        if (content.contains("/by")) {
```

Comment:
Magic strings here (`"/by"`). You might want to convert this to a named constant. There are other similar magic strings/literals/numbers issue in this and other files too. Sometimes, you could also use enum to resolve the issue. Do look out for them!